### ISSUE TITLE
ResourceInputStream stuck in same coroutine regression

### ISSUE THREAD

**Issue by ostrolucky** [amphp/byte-stream#47]
Since a4739c8a6dd96d772cfe49a95d5dfda8c7417595, following hangs with no output

```php
<?php

use Amp\ByteStream\ResourceInputStream;
use Amp\ByteStream\ResourceOutputStream;
use Amp\Loop;

require __DIR__ . "/../vendor/autoload.php";

$writeCoRoutine2 = \Amp\asyncCoroutine(function() {
    $middleReadStream = new ResourceInputStream(fopen('middle', 'r'));

    while (true) {
        echo yield $middleReadStream->read();
    }
});

Loop::run(function () use ($writeCoRoutine2) {
    \Amp\ByteStream\pipe(
        new ResourceInputStream(fopen('/home/g9735/Downloads/somefile.avi', 'r')),
        new ResourceOutputStream(fopen('middle', 'w'))
    );

    $writeCoRoutine2();
});
```

as soon as `read()` in $writeCoRoutine2 is executed, event loop will never return to pipe

**Comment by kelunik** [comment]
The actual issue is the early return in case the stream isn't readable anymore. Currently we assume a stream is closed once it's closed the first time.

It just worked before, because the first read was deferred any there was always data to read after that.

https://github.com/amphp/byte-stream/blob/3a4e493d3922198eef29d11365e72a5b8d9bc2f7/lib/ResourceInputStream.php#L121

**Comment by kelunik** [comment]
It doesn't, because of your `while (true)` loop in combination with `Success`, which is an immediately resolved promise.

**Comment by ostrolucky** [comment]
Indeed. Still two problems though. But yes, both of them would be solved by simply removing the assumption that there are not going to be more writes to resource after it's eof.

I noticed you have added a test where you solved this issue by adding Delay. That's not ideal. Now try to modify test case also like following, which should reflect my real use case more. I cannot assume same amount of read/writes will be called during piping and while loop. I might not even know the size of source since it might not be a file.

```diff
--- test/ResourceStreamTest.php	(revision aa2ebfe17be166e197375020483fb78f9d497fa6)
+++ test/ResourceStreamTest.php	(date 1545781598000)
@@ -260,7 +260,7 @@
 
             \Amp\ByteStream\pipe(
                 new ResourceInputStream(\fopen(__FILE__, 'rb')),
-                new ResourceOutputStream(\fopen($middle, 'wb'))
+                new ResourceOutputStream(\fopen($middle, 'wb'), 1)
             );
 
             $middleReadStream = new ResourceInputStream(\fopen($middle, 'rb'));
```

**Comment by ostrolucky** [comment]
ping @kelunik

**Comment by kelunik** [comment]
@ostrolucky We can't simply remove that assumption, it's how we defined `InputStream` to work. I'd advise you to work with another API or write your own stream, that never returns `null` in the first place.

Not sure whether you know it, but file resources (e.g. opened with `fopen`) will always do blocking reads, never non-blocking reads, because async I/O isn't supported for file handles.

I'm fine with a scheduling change that sometimes does a `Loop::delay(0)` / `Loop::defer` to make the scheduling fairer.

**Comment by ostrolucky** [comment]
> file resources (e.g. opened with fopen) will always do blocking reads

Really? http://php.net/manual/en/function.stream-set-blocking.php says 

>  This function works for any stream that supports non-blocking mode (currently, regular files and socket streams). 


Anyway what I intend to do when solving this issue is to get rid of amphp/file and just use byte-stream, since according my testing it results in **considerate** boost of read/write speed

**Comment by kelunik** [comment]
@ostrolucky Yes, really. While the read might return `EAGAIN` then instead of blocking, the actual problem is that polling doesn't work, it always returns immediately, see https://www.remlab.net/op/nonblock.shtml

Have you tried installing `ext-eio` or `ext-uv`?

**Comment by bwoebi** [comment]
@ostrolucky If you want raw speed, then yes, don't use amphp/file. If you want low _overall_ latency, then by all means go with amphp/file.
amphp/file is slower _by design_ in that it defers work to another thread/process. But while the other thread/process does work on executing your disk I/O, your current process can still do other work.

amphp/file is built to scale, not to be extremely fast, basically.

**Comment by kelunik** [comment]
@trowski What's your thought on the scheduling?

**Comment by trowski** [comment]
I think the scheduling should remain as-is. There's a few things wrong here that will cause issues no matter what kind of scheduling we do.

The code originally posted is performing reads with no condition. In synchronous PHP code, it's effectively doing this:

```php
$file = fopen('middle', 'r');
while (true) {
    echo fread($file, 8192);
}
```
`while (true)` with no escape condition *should* result in an infinite loop. Adding delays/defers just hides the problem IMO.

`ResourceInputStream` and `ResourceOutputStream` were never meant for files (we should probably put this as a warning box on the front page of the docs). The correct way to do this is with `amphp/file`. Install `ext-eio` or `ext-uv` if you're concerned about performance. I would also advise against using `pipe` if you want to do something else with the data read from the input stream.

**Comment by ostrolucky** [comment]
> ext-eio or ext-uv

- I'm building simple _tool_, not a server running 24/7. I don't want to make consumers of my tool install non standard php extension. Making it work with good performance should be as simple as doing `composer require` and executing it. And don't get me even started on cross-platform compatibility issues
- I did try them, they have various bugs making them unusable for me and since I don't really want to use them anyway because of first point, I don't want to make investment in creating detailed bug reports

@bwoebi That's exactly it. I don't need to scale much and hence am interested more in raw speed. I'm fine to block the script for few ms while `fread` is being called, since there is usually only 2 tasks running concurrently anyway. 

If ResourceInputStream really provides 0 advantages over just simple `fopen` + `fread` calls, I guess I am just going to drop it and do just that.

**Comment by trowski** [comment]
`InputStream` returns null from `read()` to indicate EOF. This is no longer something can or should be changed. I and many others have written libraries and apps using this API without issue. `ResourceInputStream` was designed to work with pipes and sockets, *not* files. If your script was allowed to continue to the `fread()` and subsequent watcher activation, it would result in 100% CPU as the readable watcher would immediately report the stream as readable (again, because loop watchers do not work properly for file streams). If you were using anything other than `NativeDriver` for a loop, I believe even creating the watcher would fail.

As I said before, you shouldn't use `pipe()` if you want to do something with the data read from the source. Against my better judgement, here's some code showing what I meant using file streams with `Resource*Stream` objects:

```php
Loop::run(function () {
    $source = new ResourceInputStream(\fopen('source/file/path', 'rb'));
    $destination = new ResourceOutputStream(\fopen('destination/file/path', 'wb'));

    while (($chunk = yield $source->read()) !== null) {
        echo $chunk; // Or whatever else you wish to do with $chunk.
        yield $destination->write($chunk);
    }
});
```

I'm certain you could have written that on your own, so I'm confused at your refusal to do so.

---

I would recommend a different approach to concurrency. Use `amphp/parallel` and `Amp\Parallel\Worker\Task` to download each file within a worker. You can even use `amphp/file`, which uses blocking operations when within a worker. If worker is too simple or opinionated of an API, you can use `Amp\Parallel\Context\Process` to write your own process while being provided a communication channel back to the parent. Initial file downloads/uploads will add some overhead of starting another process, but file operations within the process will not be encumbered by IPC.

**Comment by ostrolucky** [comment]
My example code is simplest reproducer. There is a reason there are 3 resources there - I don't just echo in real use case, but write to resource 4 which doesn't exist at the time piping between resource 1 and 2 is initiated

> Use amphp/parallel and Amp\Parallel\Worker\Task to download each file within a worker

Issue isn't with piping but reading incomplete file. Also, resource 4 is socket, which I believe makes `Process` suggestion also not applicable.

> You can even use amphp/file

I do, it's way slower. That's why I want to use ResourceInputStream instead and submitted this issue. Would be much faster in my use case to get rid of amp/parallel overhead and just not use 100% non-blocking code.

Feel free to experiment yourself with https://github.com/ostrolucky/stdinho and commands `head -c 1G </dev/urandom | bin/stdinho 127.0.0.1:1337 -v` and `curl 127.0.0.1:1337 > /dev/null`. Then try to replace `amphp/file` with `ResourceInputStream` inside `\Ostrolucky\Stdinho\Responder` class. Or don't, but then don't try to workaround my simple reproducer instead of solving root of the problem.

>  it would result in 100% CPU as the readable watcher would immediately report the stream as readable

It's nearly 100% even with amphp/file
